lib/authz/service.go

```diff
@@ -0,0 +1,79 @@
+package authz
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/authz/opa_files"
+	"github.com/enfabrica/enkit/lib/authz/plugins"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+)
+
+type regoInput struct {
+	Username string   `json:"username"`
+	Resource Resource `json:"resource"`
+	Action   Action   `json:"action"`
+}
+
+func (i regoInput) JSON() string {
+	js, _ := json.Marshal(i)
+	return string(js)
+}
+
+type AuthService struct {
+	q rego.PreparedEvalQuery
+}
+
+var (
+	ErrorOnlyAccessSelfOwned = errors.New("permission denied to perform this action")
+)
+
+func (as *AuthService) Do(ctx context.Context, req *AuthRequest) error {
+	i := &regoInput{
+		Username: req.UserName,
+		Resource: req.Resource,
+		Action:   req.Action,
+	}
+	res, err := as.q.Eval(ctx, rego.EvalInput(i))
+	if err != nil {
+		return err
+	}
+	fmt.Println(res)
+	if len(res) == 0 {
+		return ErrorOnlyAccessSelfOwned
+	}
+
+	if m, ok := res[0].Bindings["x"].(map[string]interface{}); ok {
```

Comment:
I would invert this if for readability:
```
if m, ok := res[0].Bindings["x"].(map[string]interface{}); !ok {
  return errors.New("wrong response type given")
}

canDo, ok := m["allow"].(bool)
if !ok {
  return errors.new("no allow statement was provided")
}

if canDo {
  return nil
}

return ErrorOnlyAccessSelfOwned
```

Reply:
BTW, this is pretty obscure... what is "x"? why res[0]? It seems to be very error prone fragile, which is a bit concerning for security code.

Reply:
is there a res[1] and res[2] to parse? do we ignore them? Can you add a comment explaining the logic/response expected? are there other keys but x we need to care about?